pkg/operator/resource/resourceapply/unstructured.go

```diff
@@ -0,0 +1,91 @@
+package resourceapply
+
+import (
+	"context"
+
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourcemerge"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog"
+)
+
+// applyUnstructured ensures the form of the specified usntructured is present in the API.
+// If it does not exist, it will be created. If it does exist, the metadata of the required
+// usntructured will be merged with the existing usntructured and an update performed if the
+// usntructured spec and metadata differ from the previously required spec and metadata.
+// For further detail, check the top-level comment.
+func applyUnstructured(resourceClient dynamic.ResourceInterface, path string, recorder events.Recorder,
```

Comment:
doesn't this ask for hot loops? Apply in a generic way does not really make sense IMO.

Reply:
So you think this needs an option to `forceApply`, same as it is for Deployments and DaemonSets? Or it should repeat the attempt couple of times before failing?

Reply:
@Danil-Grigorev Can you please build this into the MAO and try it on a real cluster to see how it behaves?

I don't think it should hot loop, it's based on the `ApplyDeployment` logic, so if that doesn't hot loop I think this shouldn't? If the metadata and spec that we input doesn't change, then the spec hash annotation shouldn't change and we should short circuit on line 42 before we attempt to update again. @sttts Is there something in this in particular that makes you think it would hot loop?

Reply:
This func looks generic. But it is wrong in most cases. It does not consider things like list-type markers (compare https://github.com/kubernetes/enhancements/blob/6a1ac61fbfd64d49e13fae3a8d773dd4ebda53c8/keps/sig-api-machinery/0006-apply.md). It might be fine for one specific resource. But having this helper here encourages simple, but wrong code.

Reply:
Once this PR passes e2e tests, we would have the proof that hot loops do not occur in current implementation: https://github.com/openshift/machine-api-operator/pull/642

Reply:
@Danil-Grigorev please check https://github.com/openshift/library-go/pull/836#issuecomment-658822346. This is a blocker for the PR.